Add authentication

andres-vidal opened the pull request:
Closes #29 

Refer to the [branch's README](https://github.com/wyeworks/richardburton/tree/add-authentication) for the specification of the implemented flow.

iaguirre88 commented on backend/lib/richard_burton/auth_google.ex:
```diff
@@ -0,0 +1,74 @@
+        )
+
+      _ ->
+        throw("Auth configuration is not properly set.s")
```
```suggestion
        throw("Auth configuration is not properly set.")
```

iaguirre88 commented on backend/lib/richard_burton/auth_google.ex:
```diff
@@ -0,0 +1,74 @@
+  def verify(token) do
+    case Application.get_env(:richard_burton, :auth_config) do
+      {%{"issuer" => issuer}, keys} ->
+        verify(
```
NIT: even though `verify/1` and `verify/4` are different functions, I think it's more idiomatic to name the private one something like `do_verify` so it's clearer for the reader which one is the public one.

iaguirre88 commented on backend/test/richard_burton/user_test.exs:
```diff
@@ -0,0 +1,79 @@
+      refute change_valid(%{"subject_id" => nil}).valid?
+    end
+
+    test "when an author with the provided attributes exist, is invalid" do
```
> when a **user** with ...

iaguirre88 commented on backend/test/richard_burton_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,42 @@
+  import Routes, only: [user_path: 2]
+
+  @valid_attrs %{"email" => "example@gmail.com"}
+  @valid_return %{"email" => "example@gmail.com", "role" => "reader"}
```
NIT: maybe `@expected_return`?

iaguirre88 commented on backend/test/richard_burton_web/controllers/user_controller_test.exs:
```diff
@@ -0,0 +1,42 @@
+  @valid_return %{"email" => "example@gmail.com", "role" => "reader"}
+
+  describe "POST /users" do
+    test "when providing valid params, returns 201 created and the inserted user", meta do
```
NIT:
```suggestion
    test "when providing valid params, returns 201 created and the inserted user", %{conn: conn} do
```

iaguirre88 reviewed (commented):
I just left a first round of comments.

I've noticed some changes related to `Author`, `OriginalBook`, `Publication`, and `TranslatedBook`. Were those changes introduced here? Or is it because the branch needs rebasing?

Let me know so I can decide whether I should review them or not.

andres-vidal commented:
@iaguirre88 indeed, the branch needs rebasing 😅

andres-vidal commented:
@iaguirre88 there it goes, addressed all your comments. LMK what you think, specially about the rewrite on the README. Thank's for carefully reading it 😊
